blockchain/peer_test.go

```diff
@@ -0,0 +1,241 @@
+	}
+	peer.resetMonitor()
+	assert.NotNil(t, peer.recvMonitor)
+}
+
+func TestPeerTimer(t *testing.T) {
+	peerTimeout = 2 * time.Millisecond
+
+	peer := &bpPeer{
+		id:      p2p.ID(cmn.RandStr(12)),
+		height:  10,
+		logger:  log.TestingLogger(),
+		errFunc: errFunc,
+	}
+	assert.Nil(t, peer.timeout)
+
+	// initial reset call with peer having a nil timer
+	peer.resetTimeout()
+	assert.NotNil(t, peer.timeout)
+	// make sure timer is running and stop it
+	checkByStoppingPeerTimer(t, peer, true)
+
+	// reset with non nil expired timer
+	peer.resetTimeout()
+	assert.NotNil(t, peer.timeout)
+	// make sure timer is running and stop it
+	checkByStoppingPeerTimer(t, peer, true)
+	resetErrors()
+
+	// reset with running timer (started above)
+	time.Sleep(time.Millisecond)
+	peer.resetTimeout()
+	assert.NotNil(t, peer.timeout)
+
+	// let the timer expire and ...
+	time.Sleep(3 * time.Millisecond)
+	checkByStoppingPeerTimer(t, peer, false)
+
+	peerTestMtx.Lock()
+	// ... check an error has been sent, error is peerNonResponsive
+	assert.Equal(t, 1, numErrFuncCalls)
+	assert.Equal(t, lastErr, errNoPeerResponse)
+	peerTestMtx.Unlock()
+
+	// Restore the peerTimeout to its original value
+	peerTimeout = 15 * time.Second
+}
+
+func TestPeerIncrPending(t *testing.T) {
+	peerTimeout = 2 * time.Millisecond
+
+	peer := &bpPeer{
+		id:      p2p.ID(cmn.RandStr(12)),
+		height:  10,
+		logger:  log.TestingLogger(),
+		errFunc: errFunc,
+	}
+
+	peer.incrPending()
+	assert.NotNil(t, peer.recvMonitor)
+	assert.NotNil(t, peer.timeout)
+	assert.Equal(t, int32(1), peer.numPending)
+
+	peer.incrPending()
+	assert.NotNil(t, peer.recvMonitor)
+	assert.NotNil(t, peer.timeout)
+	assert.Equal(t, int32(2), peer.numPending)
+
+	// Restore the peerTimeout to its original value
+	peerTimeout = 15 * time.Second
+}
+
+func TestPeerDecrPending(t *testing.T) {
+	peerTimeout = 2 * time.Millisecond
+
+	peer := &bpPeer{
+		id:      p2p.ID(cmn.RandStr(12)),
+		height:  10,
+		logger:  log.TestingLogger(),
+		errFunc: errFunc,
+	}
+
+	// panic if numPending is 0 and try to decrement it
+	assert.Panics(t, func() { peer.decrPending(10) })
+
+	// decrement to zero
+	peer.incrPending()
+	peer.decrPending(10)
+	assert.Equal(t, int32(0), peer.numPending)
+	// make sure timer is not running
+	checkByStoppingPeerTimer(t, peer, false)
+
+	// decrement to non zero
+	peer.incrPending()
+	peer.incrPending()
+	peer.decrPending(10)
+	assert.Equal(t, int32(1), peer.numPending)
+	// make sure timer is running and stop it
+	checkByStoppingPeerTimer(t, peer, true)
+
+	// Restore the peerTimeout to its original value
+	peerTimeout = 15 * time.Second
+}
+
+func TestPeerCanBeRemovedDueToExpiration(t *testing.T) {
+	minRecvRate = int64(100) // 100 bytes/sec exponential moving average
+
+	peer := &bpPeer{
+		id:      p2p.ID(cmn.RandStr(12)),
+		height:  10,
+		errFunc: errFunc,
+		logger:  log.TestingLogger(),
+	}
+
+	peerTimeout = time.Millisecond
+	peer.incrPending()
+	time.Sleep(2 * time.Millisecond)
```

Comment:
Could we have non deterministic test failure in this case?

Reply:
Shouldn't IMO